opacus/grad_sample/README.md

```diff
@@ -1,16 +1,56 @@
 
-```python
-from opacus.grad_sample import register_grad_sampler
+Each of the two implementations comes with it's own set of limitations, and we leave the choice up to the client 
+which one to use. 
 
-@register_grad_sampler(nn.MyCustomClass)
-def compute_grad_sample(module, activations, backprops):
-    pass
-```
+``GradSampleModuleExpandedWeights`` is currently in early beta and can produce unexpected errors, but potentially
+improves upon ``GradSampleModule`` on performance and functionality.
 
-Note that you can also pass a list to the decorator, and register one function against multiple nn.Module classes.
+**TL;DR:** If you want stable implementation, use ``GradSampleModule`` (`grad_sample_mode="hooks"`). 
+If you want to experiment with the new functionality - try ``GradSampleModuleExpandedWeights``(`grad_sample_mode="ew"`) 
+and switch back to ``GradSampleModule`` if you encounter strange errors or unexpexted behaviour. 
+We'd also appreciate it if you report these to us
+
+## Hooks-based approach
+- Model wrapping class: ``opacus.grad_sample.grad_sample_module.GradSampleModule``
+- Keyword argument for ``PrivacyEngine.make_private()``: `grad_sample_mode="hooks"`
+
+Computes per-sample gradients for a model using backward hooks. It requires custom grad sampler methods for every 
+trainable layer in the model. We provide such methods for most popular PyTorch layers. Additionally, client can
+provide their own grad sampler for any new unsupported layer (see [tutorial](https://github.com/pytorch/opacus/blob/main/tutorials/guide_to_grad_sampler.ipynb))
+
+## ExpandedWeigths approach
+- Model wrapping class: ``opacus.grad_sample.gsm_exp_weights.GradSampleModuleExpandedWeights``
+- Keyword argument for ``PrivacyEngine.make_private()``: `grad_sample_mode="ew"`
+
+Computes per-sample gradients for a model using core functionality available in PyTorch 1.12+. Unlike hooks-based
+grad sampler, which works on a module level, ExpandedWeights work on the function level, i.e. if your layer is not
+explicitly supported, but only uses known operations, ExpandedWeights will support it out of the box. 
+
+At the time of writing, the coverage for custom grad samplers between ``GradSampleModule`` and ``GradSampleModuleExpandedWeights``
+is roughly the same.
+
+## Comparative analysis
+
+Please note that these are known limitations and we plan to improve Expanded Weights and bridge the gap in feature completeness
+
+
+| xxx | Hooks | Expanded Weights |
+|:-----:|:-------:|:------------------:|
+| Required PyTorch version | 1.8+ | 1.12+ |
+| Development status | Underlying mechanism deprecated | Beta |
+| Performance | - | ✅ Likely up to 2.5x faster |
+| torchscript models | Not supported | ✅ Supported |
+| Client-provided grad sampler | ✅ Supported | Not supported |
+| `batch_first=False` | ✅ Supported | Not supported |
+| Most popular nn.* layers | ✅ Supported | ✅ Supported |
+| Recurrent networks | ✅ Supported | Not supported |
```

Comment:
Out of curiosity, what is the problem with using our "DPRNN" with ExpandedWeights? I'm guessing that they forbid accumulation of grad_sample?

Reply:
I'll give you a quote from @samdow which I do not claim to fully understand:

>Backstory here: the reason RNNs are difficult for a lot of people is because the kernel will also return a buffer that's used by cuda. This means we can't write a custom per sample gradients formula because we don't have any information as to what is in the buffer and what the kernels are using it for

> So this makes it weird for a lot of systems to trace because they're returning this buffer object that should never really be user facing. What we hope to do is add a flag to the RNN module that allows a user to choose to basically decompose the RNN into a linear layer (like what Opacus has). This will make it easier to have interpretable traces and let researchers experiment on something that they understand (instead of something that returns a buffer)

Reply:
Well, so it doesn't really say what would happen with a decomposed RNN like Opacus has. Let's try!

Reply:
huh, good point

Reply:
So yeah, our custom DPRNN doesn't work too. If fails in a funny way - EW just doesn't compute `grad_sample` without throwing any exceptions.
This is clearly not a desired behaviour and I've included that into a [collab](https://colab.research.google.com/drive/1ihar3_rc_P5otxRL6xP929BxjuiZRi7A#scrollTo=RdiMJNRoWKsm) documenting required improvements

Reply:
I can shine some light on this! So the more generic mechanism that EW is using is called reparametrization and it assumes that the names of the parameters match what PyTorch would create. At a high level, because the DPRNN layers are renaming their parameters, it terminates the reparameterization algorithm early so we skip turning the original weights into expanded weights (meaning that they don't get the grad_sample attribute, yikes!). This is probably going to be difficult to error on because it's hard to detect when a user is changing the names of parameters.

The reason that changing the RNN behavior with a flag should work is because it will still be the RNN layer, we'll just change what's happening in the forward pass if the flag is set. I assumed that the renaming trick was so that exporting models worked correctly so that we couldn't this behavior in Opacus, but without the renaming it should work with EW (yet not with exporting 😞 )